File: CPP/Tests/TestLines.cpp
#include <gtest/gtest.h>
#include "clipper2/clipper.h"
#include "ClipFileLoad.h"
TEST(Clipper2Tests, TestMultipleLines) {
  std::ifstream ifs("Lines.txt");
  ASSERT_TRUE(ifs.good());
  int test_number = 1;
  while (true)
  {
    Clipper2Lib::Paths64 subject, subject_open, clip;
    Clipper2Lib::Paths64 solution, solution_open;
    Clipper2Lib::ClipType ct;
    Clipper2Lib::FillRule fr;
    int64_t area, count;
    if (!LoadTestNum(ifs, test_number,
      subject, subject_open, clip, area, count, ct, fr)) break;
    Clipper2Lib::Clipper64 c;
    c.AddSubject(subject);
    c.AddOpenSubject(subject_open);
    c.AddClip(clip);
    EXPECT_TRUE(c.Execute(ct, fr, solution, solution_open));
    const int64_t count2 = solution.size() + solution_open.size();
    const int64_t count_diff = std::abs(count2 - count);
    const double relative_count_diff = count ?
      count_diff / static_cast<double>(count) :
      0;
    if (test_number == 1)
    {
      EXPECT_EQ(solution.size(), 1);
      if (solution.size() > 0)
      {
        EXPECT_EQ(solution[0].size(), 6);
        EXPECT_TRUE(IsPositive(solution[0]));
      }
      EXPECT_EQ(solution_open.size(), 1);
      if (solution_open.size() > 0)
      {
        EXPECT_EQ(solution_open[0].size(), 2);
        if (solution_open[0].size() > 0)
        {
          //expect vertex closest to input path's start
          EXPECT_EQ(solution_open[0][0].y, 6);
        }
      }
    }
    else
    {
      EXPECT_LE(count_diff, 8);
      EXPECT_LE(relative_count_diff, 0.1);
    }
    ++test_number;
  }
  ifs.close();
  EXPECT_GE(test_number, 17);
}

File: CSharp/Tests/Tests1/Tests/TestLines.cs
using Microsoft.VisualStudio.TestTools.UnitTesting;

namespace Clipper2Lib.UnitTests
{

  [TestClass]
  public class TestLines
  {

    [TestMethod]
    public void TestOpenPaths()
    {
      for (int i = 0; i <= 16; i++)
      {
        Clipper64 c64 = new();
        Paths64 subj = new(), subj_open = new(), clip = new();
        Paths64 solution = new(), solution_open = new();

        Assert.IsTrue(ClipperFileIO.LoadTestNum("..\\..\\..\\..\\..\\..\\Tests\\Lines.txt",
          i, subj, subj_open, clip, out ClipType clipType, out FillRule fillrule,  
          out long area, out int count, out _),
            string.Format("Loading test {0} failed.", i));

        c64.AddSubject(subj);
        c64.AddOpenSubject(subj_open);
        c64.AddClip(clip);
        c64.Execute(clipType, fillrule, solution, solution_open);

        if (area > 0)
        {
          double area2 = Clipper.Area(solution);
          double a = area / area2;
          Assert.IsTrue(a > 0.995 && a < 1.005,
            string.Format("Incorrect area in test {0}", i));
        }

        if (count > 0 && Math.Abs(solution.Count - count) > 0)
        {
          Assert.IsTrue(Math.Abs(solution.Count - count) < 2,
            string.Format("Incorrect count in test {0}", i));
        }

      } //bottom of num loop

    }
  }
}
